empty return statement should be a compiler error when return type is specified

Currently this function compiles just fine

```.ts
function foo(x): string {
	return;
}
```

I understand that it's happening because `undefined` is a valid result for a variable of type `string`, and this sort of related back to #185, but a midway point could be to simply require the return statement to not be empty.

Similarly, this code fails to compile:

```.ts
function foo(x): string {
}
```

But this one compiles:

```.ts
function foo(x): string {
	if(false) {
		return;
	}
}
```

I think the expected behavior is that every branch should require an explicit return statement.


A side benefit of this is that you can get exhausting matching completeness basically for free:

```.ts
function foo(x: number | string | boolean): string {
	if(typeof x === 'number') {
		return x.toString();
	} else if (typeof x === 'string') {
		return x;
	}
}
```

Currently the above compiles, but if TypeScript required an explicit return in every branch for non void return functions, there would be a compiler error because there are remaining types in the union. However this would compile:

```.ts
function foo(x: number | string | boolean): string {
	if(typeof x === 'number') {
		return x.toString();
	} else if (typeof x === 'string') {
		return x;
	} else if (typeof x === 'boolean') {
		return ''+x;
	}
}
```

This is pretty nice! There is one edge case still -- because of the implicit `void` included in `x`, calling `foo(null)` or `foo(undefined)` will return `undefined`. I don't know what to do about this because it relates pretty heavily to #185. I think for now it can be ignored, and whatever resolution that (hopefully) happens there will take care of it.

```ts
function foo(x: number | string | boolean): string {
    if(typeof x === 'number') {
        return x.toString();
    } else if (typeof x === 'string') {
        return x;
    }
}
```
as a note, currently if you compile this code snippet with `--noImplicitReturns` flag you'll get error
```
test.ts(1,45): error TS7030: Not all code paths return a value.
```

--- a/src/compiler/checker.ts
+++ b/src/compiler/checker.ts
@@ -13842,11 +13842,11 @@ namespace ts {
                 }
             }
 
-            if (node.expression) {
-                const func = getContainingFunction(node);
-                if (func) {
-                    const signature = getSignatureFromDeclaration(func);
-                    const returnType = getReturnTypeOfSignature(signature);
+            const func = getContainingFunction(node);
+            if (func) {
+                const signature = getSignatureFromDeclaration(func);
+                const returnType = getReturnTypeOfSignature(signature);
+                if (node.expression) {
                     const exprType = checkExpressionCached(node.expression);
 
                     if (func.asteriskToken) {
@@ -13880,6 +13880,10 @@ namespace ts {
                             checkTypeAssignableTo(exprType, returnType, node.expression);
                         }
                     }
+                }
+                else if (compilerOptions.noImplicitReturns && !maybeTypeOfKind(returnType, TypeFlags.Void | TypeFlags.Any)) {
+                    // The function has a return type, but the return statement doesn't have an expression.
+                    error(node, Diagnostics.Not_all_code_paths_return_a_value);
                 }
             }
         }

--- a/tests/baselines/reference/noImplicitReturnsWithoutReturnExpression.errors.txt
+++ b/tests/baselines/reference/noImplicitReturnsWithoutReturnExpression.errors.txt
@@ -1,0 +1,34 @@
+tests/cases/compiler/noImplicitReturnsWithoutReturnExpression.ts(2,5): error TS7030: Not all code paths return a value.
+tests/cases/compiler/noImplicitReturnsWithoutReturnExpression.ts(22,9): error TS7030: Not all code paths return a value.
+
+
+==== tests/cases/compiler/noImplicitReturnsWithoutReturnExpression.ts (2 errors) ====
+    function isMissingReturnExpression(): number {
+        return;
+        ~~~~~~~
+!!! error TS7030: Not all code paths return a value.
+    }
+    
+    function isMissingReturnExpression2(): any {
+        return;
+    }
+    
+    function isMissingReturnExpression3(): number|void {
+        return;
+    }
+    
+    function isMissingReturnExpression4(): void {
+        return;
+    }
+    
+    function isMissingReturnExpression5(x) {
+        if (x) {
+            return 0;
+        }
+        else {
+            return;
+            ~~~~~~~
+!!! error TS7030: Not all code paths return a value.
+        }
+    }
+    

--- a/tests/baselines/reference/noImplicitReturnsWithoutReturnExpression.js
+++ b/tests/baselines/reference/noImplicitReturnsWithoutReturnExpression.js
@@ -1,0 +1,48 @@
+//// [noImplicitReturnsWithoutReturnExpression.ts]
+function isMissingReturnExpression(): number {
+    return;
+}
+
+function isMissingReturnExpression2(): any {
+    return;
+}
+
+function isMissingReturnExpression3(): number|void {
+    return;
+}
+
+function isMissingReturnExpression4(): void {
+    return;
+}
+
+function isMissingReturnExpression5(x) {
+    if (x) {
+        return 0;
+    }
+    else {
+        return;
+    }
+}
+
+
+//// [noImplicitReturnsWithoutReturnExpression.js]
+function isMissingReturnExpression() {
+    return;
+}
+function isMissingReturnExpression2() {
+    return;
+}
+function isMissingReturnExpression3() {
+    return;
+}
+function isMissingReturnExpression4() {
+    return;
+}
+function isMissingReturnExpression5(x) {
+    if (x) {
+        return 0;
+    }
+    else {
+        return;
+    }
+}

--- a/tests/cases/compiler/noImplicitReturnsWithoutReturnExpression.ts
+++ b/tests/cases/compiler/noImplicitReturnsWithoutReturnExpression.ts
@@ -1,0 +1,25 @@
+// @noImplicitReturns: true
+function isMissingReturnExpression(): number {
+    return;
+}
+
+function isMissingReturnExpression2(): any {
+    return;
+}
+
+function isMissingReturnExpression3(): number|void {
+    return;
+}
+
+function isMissingReturnExpression4(): void {
+    return;
+}
+
+function isMissingReturnExpression5(x) {
+    if (x) {
+        return 0;
+    }
+    else {
+        return;
+    }
+}

